Update: add shadowed variable loc to message in no-shadow (fixes #13646)

t-mangoe opened the pull request:
…646)

<!--
    Thank you for contributing!

    ESLint adheres to the [JS Foundation Code of Conduct](https://js.foundation/community/code-of-conduct).
-->

#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/master/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[X] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->
#13646 

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/developer-guide/contributing/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

#### What changes did you make? (Give an overview)
I changed the no-shadow rule to show where where the original identifier has been defined.


#### Is there anything you'd like reviewers to focus on?
I am not familiar with JavaScript, so please check whether my codes have a problem.

mdjermanovic commented:
Hi @t-mangoe, thanks for the PR!

The original issue #13646 isn't accepted yet. Per our [process](https://eslint.org/docs/maintainer-guide/issues#accepting-issues), it needs two more :+1: s from team members, so we'll wait to see if it will reach consensus before reviewing.

t-mangoe commented:
I see. I am waiting.

btmills commented:
#13646 is accepted, so I've labeled this as accepted as well.

btmills commented on lib/rules/no-shadow.js:
```diff
@@ -117,6 +117,29 @@ module.exports = {
+            } else {
+                obj = {
+                    line: 0,
+                    column: 0
```
For global variables that have no location within the file to point to, pointing to  `0:0` seems a bit strange. What do you think of adding a second message saying something like "___ is already a global variable." without location information, but the "upper scope" message does since we know where to point to?

t-mangoe replied:
Thank you for your comment! I understood.
I'll change codes to show message that the variable is a global variable, instead of no location `0:0`

btmills reviewed (changes requested):
@t-mangoe thanks for putting this PR together already! Now that the issue is accepted, let's get it finished up so we can merge it.

I left one idea to improve the message when we don't have location information.

Our contributor license agreement currently thinks you haven't signed it yet. Can you [fill it out](https://cla.js.foundation/eslint/eslint?pullRequest=13841)? If you already have, perhaps it'll catch up when you push another commit to the branch, and if not, let us know and we'll see what's going on.

btmills commented:
It looks like your latest push included a bunch of unrelated commits from our `master` branch. It looks like this pull request should just include just `Update: show the original identifier place (refs #13646)` and `Update: show message when the variable is global (refs #13646)`. Can you `git rebase --interactive` on the latest `master` to include just those two?

mdjermanovic commented:
Merge commit https://github.com/t-mangoe/eslint/commit/fdd6444ace10115a74d5d461b1f1a2adb8503323 has two parents that look identical to each other. 

@t-mangoe Did you rebase onto the latest master and then pulled from your branch? If so, `git pull` created that merge commit for the original https://github.com/t-mangoe/eslint/commit/f235f99e21678c771e9fec711b6e80e4baad7fc9 and its rebased equivalent https://github.com/t-mangoe/eslint/commit/2347d4f93ed6f1994cf5ac3cb5a8d1c41add9e50.

You can write `git rebase --onto HEAD~2 HEAD~1` to remove the merge commit (`HEAD~1`) from ancestors:

<details>

```bash
$ git status
On branch where-original-identifier
Your branch is up to date with 'origin/where-original-identifier'.

nothing to commit, working tree clean

$ git log -4
commit 6c676213e08277d89af8a85d3445ce5962df57f5 (HEAD -> where-original-identifier, origin/where-original-identifier)
Author: t-mangoe <m.takashi.xpair@gmail.com>
Date:   Sat Jan 16 11:08:51 2021 +0900

    Update: show message when the variable is global (refs #13646)

commit fdd6444ace10115a74d5d461b1f1a2adb8503323
Merge: 2347d4f93 f235f99e2
Author: t-mangoe <m.takashi.xpair@gmail.com>
Date:   Sat Jan 16 11:08:26 2021 +0900

    Merge branch 'where-original-identifier' of https://github.com/t-mangoe/eslint into where-original-identifier

commit 2347d4f93ed6f1994cf5ac3cb5a8d1c41add9e50
Author: t-mangoe <m.takashi.xpair@gmail.com>
Date:   Sat Nov 14 15:00:49 2020 +0900

    Update: show the original identifier place (refs #13646)

commit 6509705a3b8d2542d09d1c22041fe73dd0d0638f
Author: ESLint Jenkins <eslint[bot]@users.noreply.github.com>
Date:   Fri Jan 15 18:00:03 2021 -0500

    7.18.0

$ git rebase --onto HEAD~2 HEAD~1
First, rewinding head to replay your work on top of it...
Applying: Update: show message when the variable is global (refs #13646)

$ git log -3
commit ...some new hash... (HEAD -> where-original-identifier)
Author: t-mangoe <m.takashi.xpair@gmail.com>
Date:   Sat Jan 16 11:08:51 2021 +0900

    Update: show message when the variable is global (refs #13646)

commit 2347d4f93ed6f1994cf5ac3cb5a8d1c41add9e50
Author: t-mangoe <m.takashi.xpair@gmail.com>
Date:   Sat Nov 14 15:00:49 2020 +0900

    Update: show the original identifier place (refs #13646)

commit 6509705a3b8d2542d09d1c22041fe73dd0d0638f
Author: ESLint Jenkins <eslint[bot]@users.noreply.github.com>
Date:   Fri Jan 15 18:00:03 2021 -0500

    7.18.0

$ git push origin where-original-identifier -f

```

</details>

t-mangoe commented:
@btmills @mdjermanovic 
I'm sorry, I rebased onto the master branch by mistake.
I fixed commits by @mdjermanovic 's method , `git rebase --onto HEAD~2 HEAD~1`.

btmills commented on lib/rules/no-shadow.js:
```diff
@@ -44,7 +44,8 @@ module.exports = {
         messages: {
-            noShadow: "'{{name}}' is already declared in the upper scope."
+            noShadow: "'{{name}}' is already declared in the upper scope {{shadowedLine}}:{{shadowedColumn}}.",
+            noShadowGlobal: "{{name}} is already a global variable."
```
Quoting the name for consistency with the other message:

```suggestion
            noShadowGlobal: "'{{name}}' is already a global variable."
```

t-mangoe replied:
I fixed the code following the advice.

btmills commented on lib/rules/no-shadow.js:
```diff
@@ -117,6 +118,29 @@ module.exports = {
+        /**
+         * Get declared line and column of a variable.
+         * @param {eslint-scope.Variable} variable The variable to get.
+         * @returns {Object|undefined} The declared line and column of the variable.
```
Thankfully this can't ever be `undefined` because it always goes through either the `if` or `else` branch.

```suggestion
         * @returns {Object} The declared line and column of the variable.
```

t-mangoe replied:
I removed `undefined`.

btmills reviewed (commented):
Thanks for cleaning that up, much easier to review now. I'm really liking the specific messages for globals vs declared names. Two tiny tweaks then this LGTM. Nice work!

mdjermanovic commented on lib/rules/no-shadow.js:
```diff
@@ -44,7 +44,8 @@ module.exports = {
 
         messages: {
-            noShadow: "'{{name}}' is already declared in the upper scope."
+            noShadow: "'{{name}}' is already declared in the upper scope {{shadowedLine}}:{{shadowedColumn}}.",
```
Will it be clear enough that `{{shadowedLine}}:{{shadowedColumn}}` refers to a line:column location?

Should we maybe say `"on line {{shadowedLine}} column {{shadowedColumn}}"`, or maybe just `"on {{shadowedLine}}:{{shadowedColumn}}"`.

t-mangoe replied:
I changed the message to `on line {{shadowedLine}} column {{shadowedColumn}}`

J-Vin commented on lib/rules/no-shadow.js:
```diff
@@ -165,9 +189,16 @@ module.exports = {
+                    const messageId = location.global ? "noShadowGlobal" : "noShadow";
+
+                    if (!location.global) {
+                        variable.shadowedLine = location.line;
+                        variable.shadowedColumn = location.column;
```
👍

t-mangoe replied:
I made constant `data` and passed it.

J-Vin reviewed (approved):
👍

mdjermanovic reviewed (approved):
LGTM, thanks!

btmills reviewed (approved):
LGTM, thanks @t-mangoe! This'll go into tonight's release.
